modules/internal/BytesStringCommon.chpl

```diff
@@ -1243,6 +1244,115 @@ module BytesStringCommon {
     return localX[start..end];
   }
 
+  iter theseBytes(const ref x: ?t) {
+    if x.isEmpty() then return;
+    const localX = x.localize();
+    foreach i in localX.byteIndices do
+      yield localX.byte(i);
+  }
+
+  iter theseBytes(param tag: iterKind, const ref x: ?t)
+      where tag==iterKind.leader {
+
+    const localX = x.localize();
+    for rangeChunk in localX.byteIndices.these(tag) do
+      yield rangeChunk;
+  }
+
+  iter theseBytes(param tag: iterKind, const ref x: ?t, followThis)
+      where tag==iterKind.follower {
+
+    const localX = x.localize();
+
+    for i in localX.byteIndices.these(tag, followThis) do
+      yield localX.buff[i];
+  }
+
+  iter theseAscii(const ref x: ?t) {
+    if x.isEmpty() then return;
+    const localX = x.localize();
+    foreach i in localX.byteIndices {
+      var (newBuff, allocSize) = bufferCopyLocal(localX.buff+i, len=1);
+      if localX.type == string {
+        yield chpl_createStringWithOwnedBufferNV(newBuff, 1, allocSize, 1);
+      }
+      else {
+        yield createBytesWithOwnedBuffer(newBuff, 1, allocSize);
+      }
+    }
+  }
+
+  iter theseAscii(param tag: iterKind, const ref x: ?t)
+      where tag==iterKind.leader {
+
+    const localX = x.localize();
+
+    for rangeChunk in localX.byteIndices.these(tag) do
+      yield rangeChunk;
+  }
+
+  iter theseAscii(param tag: iterKind, const ref x: ?t, followThis)
+      where tag==iterKind.follower {
+
+    const localX = x.localize();
+
+    for i in localX.byteIndices.these(tag, followThis) {
+      var (newBuff, allocSize) = bufferCopyLocal(localX.buff+i, len=1);
+      if localX.type == string then
+        yield chpl_createStringWithOwnedBufferNV(newBuff, 1, allocSize, 1);
+      else
+        yield createBytesWithOwnedBuffer(newBuff, 1, allocSize);
+    }
+  }
+
+  iter theseUTF8(const ref x: ?t) {
+    const localX = x.localize();
+
+    var i = 0;
+    while i < localX.buffLen {
+      const curPos = localX.buff+i;
+      const (decodeRet, cp, nBytes) = decodeHelp(buff=localX.buff,
+                                                 buffLen=localX.buffLen,
+                                                 offset=i,
+                                                 allowEsc=true);
+      var (newBuf, newSize) = bufferCopyLocal(curPos, nBytes);
+      yield chpl_createStringWithOwnedBufferNV(newBuf, nBytes, newSize, 1);
+
+      i += nBytes;
+    }
+  }
+
+  iter theseUTF8(param tag: iterKind, const ref x: ?t)
+      where tag==iterKind.leader {
+
+    const localX = x.localize();
+
+    for followThis in x.indices.these(tag) do yield followThis;
+  }
+
+
+  iter theseUTF8(param tag: iterKind, const ref x: ?t, followThis)
+      where tag==iterKind.follower {
+
+    const localX = x.localize();
+    const charRange = followThis[0];
+
+    // walk to the starting byte
+    var startingByte = 0:byteIndex;
+    var curIdx = 0;
+    for (cp, byteIdx, nBytes) in x._cpIndexLen() {
```

Comment:
Am I understanding correctly that this is iterating over all code points from the beginning of the string until we get one that is within the indices described by `followThis`?  I realize it's effectively equivalent at the end of the day, but is there a way to "randomly access" into `charRange.low`, and then start iterating from there?  (understanding in asking that the random access would require a loop itself, but it somehow seems as though it would make the code structure clearer... if such helpers already existed (and if someone were to optimize that random access for strings with ascii prefixes, we'd get the benefit here immediately rather than having to recreate it here?)